apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsImageControl.tsx

```diff
@@ -0,0 +1,128 @@
+import wretch from "wretch"
+
+import { JSON_FORMS_RANKING } from "~/constants/formBuilder"
+import { trpc } from "~/utils/trpc"
+import {
+  IMAGE_UPLOAD_ACCEPTED_MIME_TYPES,
+  MAX_IMG_FILE_SIZE_BYTES,
+} from "./constants"
+
+export const jsonFormsImageControlTester: RankedTester = rankWith(
+  JSON_FORMS_RANKING.ImageControl,
+  and(
+    isStringControl,
+    schemaMatches((schema) => schema.format === "image"),
+  ),
+)
+export function JsonFormsImageControl({
+  data,
+  label,
+  handleChange,
+  path,
+  description,
+  required,
+}: ControlProps) {
+  const { pageId, siteId } = useParams()
+
+  const [selectedFile, setSelectedFile] = useState<File | undefined>()
+  const [pendingFile, setPendingFile] = useState<File | undefined>()
+  const [errorMessage, setErrorMessage] = useState("")
+
+  useEffect(() => {
+    if (!!data) {
+      wretch(data as string)
+        .get()
+        .blob()
+        .then((blob) => {
+          const splitData = (data as string).split("/")
+          const fileName = splitData[-1] || "Current Image"
+          setSelectedFile(new File([blob], fileName))
+        })
+        .catch((error) => {
+          console.log("error fetching initial image", error)
+        })
+    }
+    // NOTE: Using empty dependency array because we are checking if fetch is needed only upon initial load.
+    // After this load, we are the only editor of this page, and any image url changes are caused by us and we will be caching the file locally.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+  const getPresignedMutation =
+    trpc.page.getPresignUrlForImageUpload.useMutation()
+  const uploadImage = async (image: File) => {
+    const { presignedUploadURL, fileURL } =
+      await getPresignedMutation.mutateAsync({
+        pageId: Number(pageId),
+        siteId: Number(siteId),
+      })
+    const response = await wretch(presignedUploadURL)
+      .content(image.type)
+      .put(image)
+      .res()
+    if (response.ok) {
+      setSelectedFile(pendingFile)
+      setErrorMessage("")
+      handleChange(path, fileURL)
+      console.log("new file url", fileURL)
+    } else {
+      setPendingFile(undefined)
+      setErrorMessage(
+        "There is an error uploading your file. Please try again or contact support.",
+      )
+      console.log("file upload failure", response)
+    }
+  }
+
+  return (
+    <Box py={2}>
+      <FormControl isRequired={required} isInvalid={errorMessage !== ""}>
+        <FormLabel description={description}>{label}</FormLabel>
+        <Attachment
+          name="image-upload"
+          imagePreview="large"
+          multiple={false}
+          value={selectedFile}
+          onChange={(file) => {
+            console.log(file?.name)
+            if (file) {
+              setPendingFile(file)
+              void uploadImage(file)
+            } else {
+              // NOTE: Do we need to update backend on removal of file?
```

Comment:
Ideal flow might be to have an assets table with mapping

siteId assetId isDeleted

We can then cleanse our S3 bucket by walking through the isDeleted column